cadl-extension/src/main/java/com/azure/cadl/Main.java

```diff
@@ -115,28 +92,58 @@ public static void main(String[] args) throws IOException {
                 // but we continue so user can still check the file and see why format fails
             }
             formattedFiles.put(filePath, formattedSource);
         });
 
         // write output
         // java files
-        formattedFiles.forEach((filePath, formattedSource) -> cadlPlugin.writeFile(outputFolderFinal + filePath, formattedSource, null));
+        formattedFiles.forEach((filePath, formattedSource) -> cadlPlugin.writeFile(filePath, formattedSource, null));
 
         // XML include POM
-        javaPackage.getXmlFiles().forEach(xmlFile -> cadlPlugin.writeFile(outputFolderFinal + xmlFile.getFilePath(), xmlFile.getContents().toString(), null));
+        javaPackage.getXmlFiles().forEach(xmlFile -> cadlPlugin.writeFile(xmlFile.getFilePath(), xmlFile.getContents().toString(), null));
         // Others
-        javaPackage.getTextFiles().forEach(textFile -> cadlPlugin.writeFile(outputFolderFinal + textFile.getFilePath(), textFile.getContents(), null));
+        javaPackage.getTextFiles().forEach(textFile -> cadlPlugin.writeFile(textFile.getFilePath(), textFile.getContents(), null));
         // resources
         String artifactId = ClientModelUtil.getArtifactId();
         if (!CoreUtils.isNullOrEmpty(artifactId)) {
-            cadlPlugin.writeFile(outputFolderFinal + "src/main/resources/" + artifactId + ".properties",
+            cadlPlugin.writeFile("src/main/resources/" + artifactId + ".properties",
                     "name=${project.artifactId}\nversion=${project" + ".version}\n", null);
         }
     }
 
+    private static CadlPlugin.Options loadCadlOptions(String[] args, CodeModel codeModel) {
+        String outputFolder = "cadl-tests/cadl-output/";
+
+        if (args.length >= 2) {
+            outputFolder = args[1];
+            if (!outputFolder.endsWith("/")) {
+                outputFolder += "/";
+            }
+        }
+
+        final String outputFolderFinal = outputFolder;
+
+        String emitterOptionsJson = Configuration.getGlobalConfiguration().get("emitterOptions");
+        Map<String, Object> emitterOptions = new HashMap<>();
+        if (!CoreUtils.isNullOrEmpty(emitterOptionsJson)) {
+            try {
+                emitterOptions = OBJECT_MAPPER.readValue(emitterOptionsJson, Map.class);
+            } catch (JsonProcessingException e) {
+                LOGGER.info("Read emitter options failed, emitter options json: {}", emitterOptionsJson);
+            }
+        }
+        if (!emitterOptions.containsKey("namespace")) {
+            if (codeModel.getLanguage().getJava() != null && !CoreUtils.isNullOrEmpty(codeModel.getLanguage().getJava().getNamespace())) {
+                emitterOptions.put("namespace", codeModel.getLanguage().getJava().getNamespace());
+            }
+        }
+        emitterOptions.put("output-folder", outputFolderFinal);
+        return new CadlPlugin.Options().withEmitterOptions(emitterOptions);
```

Comment:
Just `setEmitterOptions` or `addEmitterOptions`?

---

Why serialize to a `Map<String, Object>` and set the generic map via `withEmitterOptions`? We prefer a model whenever we know the data.
We write to and read from this "emitterOptions", then why keep ourselves in the dark about its possible content?

Reply:
Yes, that's the part I felt weird about. My initial though was to ease the process of increasing options over time, but the map would confuse the dev about what the possible options could be. And if we want to control what options to accept, yes it's better to use a model.